test/boost/sstable_compaction_test.cc

```diff
@@ -1217,6 +1218,96 @@ SEASTAR_TEST_CASE(tombstone_purge_test) {
+            }
+            return compact_sstables(env, sstables::compaction_descriptor(to_compact), cf, sst_gen).get().new_sstables;
+        };
+
+        auto next_timestamp = [] {
+            static thread_local api::timestamp_type next = 1;
+            return next++;
+        };
+
+        auto make_insert = [&] (partition_key key, int32_t ck = 0, int32_t value = 1, std::optional<api::timestamp_type> timestamp = std::nullopt, std::optional<api::timestamp_type> created_at = std::nullopt) {
+            mutation m(s, key);
+            if (!timestamp) {
+                created_at = timestamp = next_timestamp();
+            } else if (!created_at) {
+                created_at = next_timestamp();
+            }
+            auto c_key = clustering_key::from_single_value(*s, int32_type->decompose(data_value(ck)));
+            testlog.info("make_insert: key={} ck={} timestamp={} created_at={}", dht::decorate_key(*s, key), ck, *timestamp, *created_at);
+            m.set_clustered_cell(
+                    c_key,
+                    bytes("value"),
+                    data_value(value),
+                    *timestamp);
+            m.partition().clustered_row(*s, c_key).apply(row_marker(*created_at));
+            return m;
+        };
+
+        auto make_delete_row = [&] (partition_key key, int32_t ck = 0, gc_clock::time_point deletion_time = gc_clock::now(), std::optional<api::timestamp_type> timestamp = std::nullopt, std::optional<api::timestamp_type> deleted_at = std::nullopt) {
+            mutation m(s, key);
+            if (!timestamp) {
+                deleted_at = timestamp = next_timestamp();
+            } else if (!deleted_at) {
+                deleted_at = next_timestamp();
+            }
+            shadowable_tombstone shadowable(*deleted_at, deletion_time);
+            auto c_key = clustering_key::from_single_value(*s, int32_type->decompose(data_value(ck)));
+            testlog.info("make_delete_row: key={} ck={} shadowable_tombstone={}", dht::decorate_key(*s, key), ck, shadowable);
+            m.partition().clustered_row(*s, c_key).apply(shadowable);
+            return m;
+        };
+
+        auto alpha = partition_key::from_exploded(*s, {to_bytes("alpha")});
+        auto beta = partition_key::from_exploded(*s, {to_bytes("beta")});
+
+        {
+            // Simulate materialized views update
+            // We expect mut2 to delete mut1, and be purged
+            // Since the shadowable tombstone in mut4 is ignored as it is dead
+            // and insert in mut5 has higher timestamp.
+            // This will leave only the insert in mut3 when compacting mut1-3 together.
+            auto mut1 = make_insert(alpha, 1, 1, api::timestamp_type(1), api::timestamp_type(1));
+            auto mut2 = make_delete_row(alpha, 1, gc_clock::now(), api::timestamp_type(1), api::timestamp_type(2));
+            auto mut3 = make_insert(alpha, 2, 2, api::timestamp_type(1), api::timestamp_type(2));
```

Comment:
Here, it looks like you are simulating an overwrite of the non-pk column in the base which is a pk column in the view. Shouldn't we use timestamp `1` for the shadowable tombstone? IIUC choosing `2` should be the right option if we do what proposed in issue https://github.com/scylladb/scylladb/issues/19977, but isn't that issue going to be obsolete after this PR is merged?

Reply:
What row exactly are you referring to?

Reply:
I assume the that test wanted to simulate three view updates, generated from three inserts to the base table:

```
-- Not sure if correct commands, just to illustrate:
-- CREATE TABLE base_table (id blob primary key, ck int, value int)
-- CREATE MATERIALIZED VIEW mv AS SELECT id, ck, value FROM mv WHERE id IS NOT NULL AND ck IS NOT NULL PRIMARY KEY (pk, ck);
INSERT INTO base_table (id, ck, value) VALUES ("alpha", 1, 1) USING TIMESTAMP 1;
INSERT INTO base_table (id, ck, value) VALUES ("alpha", 2, 2) USING TIMESTAMP 2;
INSERT INTO base_table (id, ck, value) VALUES ("alpha", 3, 3) USING TIMESTAMP 3;
```

The mutations I attached my first comment to should correspond to the second insert. That insert should generate two view updates, one is a shadowable tombstone _with timestamp 1_ which deletes the row generated by the first insert, and the second one inserts new data with timestamp 2. If the changes described in the linked issue were implemented, the shadowable tombstone should be issued with timestamp 2 - like you do right now - but for now it should have timestamp 1.

By the way, if the test is supposed to "Simulate materialized views update" then it would be nice to, in a comment, include the CQL commands whose effect the test was supposed to simulate. It would be easier to verify its "realness" and understand what it was supposed to simulate.

Reply:
> I assume the that test wanted to simulate three view updates, generated from three inserts to the base table:
> 
> ```
> -- Not sure if correct commands, just to illustrate:
> -- CREATE TABLE base_table (id blob primary key, ck int, value int)
> -- CREATE MATERIALIZED VIEW mv AS SELECT id, ck, value FROM mv WHERE id IS NOT NULL AND ck IS NOT NULL PRIMARY KEY (pk, ck);
> INSERT INTO base_table (id, ck, value) VALUES ("alpha", 1, 1) USING TIMESTAMP 1;
> INSERT INTO base_table (id, ck, value) VALUES ("alpha", 2, 2) USING TIMESTAMP 2;
> INSERT INTO base_table (id, ck, value) VALUES ("alpha", 3, 3) USING TIMESTAMP 3;
> ```
> 
> The mutations I attached my first comment to should correspond to the second insert. That insert should generate two view updates, one is a shadowable tombstone _with timestamp 1_ which deletes the row generated by the first insert, and the second one inserts new data with timestamp 2. If the changes described in the linked issue were implemented, the shadowable tombstone should be issued with timestamp 2 - like you do right now - but for now it should have timestamp 1.
> 
> By the way, if the test is supposed to "Simulate materialized views update" then it would be nice to, in a comment, include the CQL commands whose effect the test was supposed to simulate. It would be easier to verify its "realness" and understand what it was supposed to simulate.

I implemented the c++ unit test, before the cql-pytest, maybe it's redundant and we can just use the cql-pytest, since it tests a more real-life scenario.

Reply:
I fixed the timestamps in the next version and added comments documenting the respective cql commands.
It's a little different than the ones you guessed above, as `value` is inserted only once, using timestamp 1, and after that, only `ck` is updated to keep `value` with timestamp 1 in the row markers view updates generate.